gateway/middleware.go

```diff
@@ -682,6 +684,8 @@ func (t *BaseMiddleware) ApplyPolicies(session *user.SessionState) error {
 		session.Tags = appendIfMissing(session.Tags, tag)
 	}
 
+	session.MetaData = metadata
```

Comment:
Maps are pointers, this still leaves a data race over session.MetaData. Read at L386, Write here;

Suggestion:

- *user.SessionState needs to hold a (rw)mutex,
- protect access to underlying maps with a getter/setter,
- TT-5486 also diagnoses issues in Clone (see comment)

Ultimately the fix should be using an interface for session access/modification, as we need to protect several fields within (access rights, oauthkeys, apply policies, metadata, tags) due to the concurrency. Several functions also need pointer receivers to avoid copying the mutex and causing new issues.

How about avoiding concurrency altogether by having request scoped allocations?

Reply:
Opened https://github.com/TykTechnologies/tyk/pull/6154 ; this should prevent concurrency issues altogether, as the returned value from the session cache is a new allocation (not shared between goroutines, hopefully).